File: InfiniteScrollViewDemo/FlickrModel.m
//
//  FlickrModel.m
//  InfiniteScrollViewDemo
//
//  Created by pronebird on 12/20/16.
//  Copyright © 2016 codeispoetry.ru. All rights reserved.
//

#import "FlickrModel.h"

static NSString * const FlickrModelAttributeLink        = @"link";
static NSString * const FlickrModelAttributeMedia       = @"media";

static NSString * const FlickrMediaModelAttributeMedium = @"m";

@implementation FlickrMediaModel

- (instancetype)initWithDictionary:(NSDictionary *)dictionary {
    self = [super init];
    if(!self) {
        return nil;
    }
    
    NSString *mediumURLString = dictionary[FlickrMediaModelAttributeMedium];
    if(!mediumURLString) {
        return nil;
    }
    
    NSURL *mediumURL = [NSURL URLWithString:mediumURLString];
    if(!mediumURL) {
        return nil;
    }
    
    self.medium = mediumURL;
    
    return self;
}

@end

@implementation FlickrModel

+ (instancetype)modelWithDictionary:(NSDictionary *)dictionary {
    return [[self alloc] initWithDictionary:dictionary];
}

+ (NSArray<FlickrModel *> *)modelsFromArray:(NSArray<NSDictionary *> *)array {
    NSMutableArray *models = [[NSMutableArray alloc] init];
    
    for(NSDictionary *dict in array) {
        FlickrModel *model = [self modelWithDictionary:dict];
        if(model) {
            [models addObject:model];
        }
    }
    
    return [models copy];
}

- (id)initWithDictionary:(NSDictionary *)dictionary {
    self = [super init];
    if(!self) {
        return nil;
    }
    
    NSString *linkString = dictionary[FlickrModelAttributeLink];
    NSDictionary *mediaDictionary = dictionary[FlickrModelAttributeMedia];
    if(!linkString || !mediaDictionary) {
        return nil;
    }
    
    FlickrMediaModel *media = [[FlickrMediaModel alloc] initWithDictionary:mediaDictionary];
    NSURL *linkURL = [NSURL URLWithString:linkString];
    if(!linkURL || !media) {
        return nil;
    }
    
    self.link = linkURL;
    self.media = media;
    
    return self;
}

@end

File: InfiniteScrollViewDemoSwift/FlickrModel.swift
//
//  FlickrModel.swift
//  InfiniteScrollViewDemoSwift
//
//  Created by pronebird on 12/20/16.
//  Copyright © 2016 pronebird. All rights reserved.
//

import Foundation

struct FlickrMedia {
    let medium: URL?
}

struct FlickrResponse: Decodable {
    let link: URL?
    private let rawMedia: [String: URL]
    
    var media: FlickrMedia {
        return FlickrMedia(medium: rawMedia["m"])
    }
    
    private enum CodingKeys: String, CodingKey {
        case link, rawMedia = "media"
    }
}

private struct FlickModelAttributes {
    struct mediaAttributes {
        static let medium = "m"
    }
    
    static let media = "media"
    static let link = "link"
}

class FlickrModel {
    struct Media {
        let medium: URL
    }
    
    let link: URL
    let media: Media
    
    init?(_ dictionary: [String: Any]) {
        guard let linkUrlString = dictionary[FlickModelAttributes.link] as? String,
              let linkUrl = URL(string: linkUrlString) else { return nil }
        
        guard let mediaDictionary = dictionary[FlickModelAttributes.media] as? [String: String],
              let mediumPhotoUrlString = mediaDictionary[FlickModelAttributes.mediaAttributes.medium],
              let mediumPhotoUrl = URL(string: mediumPhotoUrlString) else { return nil }
        
        link = linkUrl
        media = Media(medium: mediumPhotoUrl)
    }
    
}
